docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,519 @@
+how the function looks up the ECALL wrapper from the `__oe_ecalls_table`.
+
+```
+...
+ecall_table.ecalls = __oe_ecalls_table;
+ecall_table.num_ecalls = __oe_ecalls_table_size;
+
+if (args.function_id >= ecall_table.num_ecalls)
+    OE_RAISE(OE_NOT_FOUND);
+
+func = ecall_table.ecalls[args.function_id];
+...
+
+// Call the function.
+func(
+    input_buffer,
+    args.input_buffer_size,
+    output_buffer,
+    args.output_buffer_size,
+    &output_bytes_written);
+```
+
+After looking up based on the function id (`sample_fcn_id_sample_ecall`),
+`oe_handle_call_enclave_function()` calls into the wrapper function `ecall_sample_ecall()`(defined in `samlpe_t.c`).
+The wrapper function performs a series of checks, unmarshals the arguments, and invokes
+the ECALL implemented by the enclave with the following code snippet.
+
+```
+/* Call user function. */
+pargs_out->_retval = sample_ecall(
+);
+```
+
+At this point, an ECALL is successfully dispatched to the enclave.
+
+**OCALL Invocation**
+
+The invocation of OCALLs is similar to that of ECALLs. The only difference is that
+the array of the function pointers is defined as `__sample_ocall_function_table`
+and is kept as part of the `enclave` struct.
+More specifically, `sample_u.c` defines a wrapper function `oe_create_sample_enclave()` as follows.
+
+```
+oe_result_t oe_create_core_enclave(
+    const char* path,
+    oe_enclave_type_t type,
+    uint32_t flags,
+    const oe_enclave_setting_t* settings,
+    uint32_t setting_count,
+    oe_enclave_t** enclave)
+{
+    return oe_create_enclave(
+               path,
+               type,
+               flags,
+               settings,
+               setting_count,
+               __sample_ocall_function_table,
+               1,
+               enclave);
+}
+```
+
+The wrapper function is used to create the enclave by the host. The `__sample_ocall_function_table`
+is used during the initialization of the `enclave` struct as follows.
+
+```
+enclave->ocalls = (const oe_ocall_func_t*)ocall_table;
+enclave->num_ocalls = ocall_count;
+```
+
+Similarly, the `oe_handle_call_host_function()` function uses these variables to handle the requests
+of OCALLs.
+
+```
+...
+ocall_table.ocalls = enclave->ocalls;
+ocall_table.num_ocalls = enclave->num_ocalls;
+
+if (args_ptr->function_id >= ocall_table.num_ocalls)
+    OE_RAISE(OE_NOT_FOUND);
+
+func = ocall_table.ocalls[args_ptr->function_id];
+...
+
+// Call the function.
+func(
+    args_ptr->input_buffer,
+    args_ptr->input_buffer_size,
+    args_ptr->output_buffer,
+    args_ptr->output_buffer_size,
+    &args_ptr->output_bytes_written);
+```
+
+Limitations
+---------
+
+The current implementation poses the following limitations.
+
+- Single enclave cannot include multiple EDL files.
+
+  According to the implementation, the `oecore` library
+directly links against the `__oe_ecalls_table` defined in a `_t.c` file.
+Therefore, the current logic cannot handle the case of including
+multiple `_t.c` files.
```

Comment:
Is there a need for one enclave including multiple `_t.c` files?

Reply:
Currently, this is only the internal use cases (e.g., `core.edl`, `syscall.edl`, etc.). Besides, I don't really see why we need to limit only a single external EDL file per enclave.

Reply:
If we are allowing multiple EDL files per enclave, then there it would make sense to revisit the `import` feature in EDL.

- When/Why would the user use the import statement ?
- When/Why would the user use multiple EDL files?

Reply:
I kind of thinking this as an option for users; i.e., they can choose either option based on their needs.

For example, users can import "string.edl" in "book.edl" and define a "math.edl" that implements an irrelevant set of ECALLs/OCALLs.

Reply:
But what would be the guideline for users? There does not seem to be a clear benefit that either approach has over the other. If I can specify multiple EDLs per enclave, then I don't see much value in allowing imports.

Reply:
An internal use case of core.edl:
```
enclave
{
    /* The end goal is to publish all EDL files as part of the SDK. At that time
     * enclave developers will simply import the edl files/functions they
     * need and this file will no longer be required. These files are imported
     * as a convenience to allow refactoring of system EDL files without the need
     * to add a new build definition for each file.
     */
    from "attestation.edl" import *;
    from "keys.edl" import *;
    from "logging.edl" import *;
    from "memory.edl" import *;
};
```

Reply:
But with this PR, we are also saying that the user can use 
"my_enclave.edl" and "attestation.edl" separately into their enclave without my_enclave.edl having to import "attestation.edl".
So it is not clear to my when I'd choose to import.

Reply:
@mingweishih core.edl and syscall.edl are helpers to make the transition to EDL import easier. We shouldn't really encourage using them.

I agree with Anand, I don't think supporting both mechanisms is beneficial.

Personally I prefer the import method because EDL already supports imports and it's easier to explain to users. From a technical standpoint, they mostly do that same thing.

Reply:
I would think from the other way:  the SDK internal use of EDLs does not entirely use import. Instead, it currently uses the `_t_wrapper.c`/`_u_wrapper.c` (and includes a table registration function that requires the oecore to link against). This refactoring that gets rid of `_t_wrapper.c`/`_u_wrapper.c` naturally supports an enclave to include multiple EDL files. I don't really think there is a reason to limit that.

This is also necessary for the `oecore` refactoring. For example, moving the attestation-related OCALLs/OCALLs out.

Reply:
@mingweishih Can you move the "multiple EDL files" per enclave requirement to next iteration of this feature? Allowing multiple EDL files per enclave makes the import feature redundant; and such a proposal would likely require multiple releases to accomplish.
If you remove that requirement, then the scope is reduced and I feel it would be easier to get consensus.

Reply:
> I would think from the other way: the SDK internal use of EDLs does not entirely use import.

@mingweishih The current plan is to require the user to explicitly import the system EDLs so that the "ghost ocalls" are explicitly opted-in.

This PR says that the user does not have to import system EDLs, rather the user can just use the system edls in the enclave without importing them. 
It is not clear what the user experience is when explicitly opting in to ghost ocalls.
Hence, my suggestion would be to amend the proposal to have minimal impact on existing features, arrive at consensus and propose a bigger change as a separate proposal.

Reply:
Here is the accepted system ocall opting in/out design doc:
https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/system_ocall_opt_in.md

Reply:
@CodeMonkeyLeet 
Yes, you can import the same edl multiple times and it behaves as if the imported item (the whole edl or a subset of it) has been imported only once.

Reply:
Registering each function by hash (with the use of `weak symbol`) effectively avoids the duplications regardless how the same ECALL is included (e.g., directly defined in an EDL or imported).  That is, the function with the same hash can only be registered once. One issue remains is that two overlapped function with the same name but different implementation. However, I think this case does not make much sense in C as opposed C++.

The complication of whole problem actually comes from the fact that we support `import` in EDL. This means that how an EDL is used is not deterministic, which is also why using per EDL table is insufficient.